packages/components/src/combobox-control/index.tsx

```diff
@@ -266,18 +270,26 @@ function ComboboxControl( props: ComboboxControlProps ) {
 
 	// Update current selections when the filter input changes.
 	useEffect( () => {
-		const hasMatchingSuggestions = matchingSuggestions.length > 0;
-		const hasSelectedMatchingSuggestions =
-			getIndexOfMatchingSuggestion(
-				selectedSuggestion,
-				matchingSuggestions
-			) > 0;
-
-		if ( hasMatchingSuggestions && ! hasSelectedMatchingSuggestions ) {
-			// If the current selection isn't present in the list of suggestions, then automatically select the first item from the list of suggestions.
-			setSelectedSuggestion( matchingSuggestions[ 0 ] );
+		if ( ! shouldFilter ) {
+			setSelectedSuggestion( null );
+			return;
 		}
-	}, [ matchingSuggestions, selectedSuggestion ] );
+
+		// If the current selection isn't present in the list of suggestions, then automatically select the first item from the list of suggestions.
+		setSelectedSuggestion( ( previousSuggetion ) => {
```

Comment:
Nice improvement, using the callback version of the state setter instead of the state variable!

As a tiny nit, there is a typo

```suggestion
		setSelectedSuggestion( ( previousSuggestion ) => {
```